[Refactor] core.common to new opensearch-common library

nknize opened the pull request:
Refactors all of o.opensearch.common classes in the core library to a new opensearch-commons library (inspired by apache commons). The intent is to refactor any o.opensearch.common.* classes in :server module to this new library. This will be done with care such that the dependencies are carefully managed to avoid any cyclic or unnecessarily complicated dependencies.

relates #5910

reta commented:
@nknize I am a bit lost now, I thought that https://github.com/opensearch-project/OpenSearch/pull/5902 concluded we don't want/need to have `opensearch-commons`, the `opensearch-core` should the one? Also, those 2 always go together, this is what every `build.gradle` now has:

```
  api project(':libs:opensearch-commons')
  api project(':libs:opensearch-core')
```

Is the `opensearch-commons` is useful by itself? 

On a slightly different subject, it seems like the convention for other modules in OpenSearch is to use `common` (not `commons`) suffix:  `analysis-common`, `ingest-common`, ...

nknize commented:
> I thought that https://github.com/opensearch-project/OpenSearch/pull/5902 concluded we don't want/need to have `opensearch-commons`

So, yes, this PR could be postponed until later. I opened it now, though, mainly to control library bloat. As I refactor the common interfaces from `:server` I was able to keep `opensearch-core` library lean by further separating the common classes into their own library for better modularity (largely in prep for jigsaw module support). That is, common classes (most in the `:server:src/o.opensearch.common` folder) refactored to this new (largely dependency free) `opensearch-commons library` under the `o.opensearch.common` namespace (like the apache commons library) and `opensearch-core` library leaned out to contain the base server classes like x-content `MediaType`, and action classes like `ActionListener`, `ActionRequest`, whose concrete implementations are in the modules. In practice, though, yes a module would depend on both `opensearch-commons` and `opensearch-core` so maybe the result of having two jars instead of one is less than ideal? I do like the security layer of explicitly defining the `exports` in `module-info.txt` for each library but we could do that in one library.

nknize commented:
If preferred I could do all this in a feature branch. Maybe that will better communicate the working backwards end state w/ less risk?

nknize commented:
Here's a [feature branch](https://github.com/nknize/OpenSearch/tree/feature/modularize) that can be tracked. I'll switch this PR to a WIP for now.

nknize commented:
@reta a top issue we face by refactoring foundation classes from the `:server` module to the `opensearch-core` library is that JarHell doesn't run on `opensearch-core` (because it defines the JarHell classes). At minimum we'll need to create a second (dependency-free) library that provides the JarHell implementation so it can be run on the core library. That was part of my thinking in creating the dependency free `commons` library, not just for precommit build classes but for any dependency-free implementation that doesn't require jar hell analysis.

reta commented:
> At minimum we'll need to create a second (dependency-free) library that provides the JarHell implementation so it can be run on the core library. T

@nknize it is certainly good idea, I second it

nknize commented:
> it is certainly good idea, I second it

What do you think about doing this with `opensearch-common` (dependency free, contains JDKJarHell) and `opensearch-core` libraries as done in this PR? I also lean toward this PR implementation where the `common` package lives in `opensearch-common` (since it's dependency free). However, I can see an argument moving it to core so it can be jarhell analyzed.

reta commented:
> What do you think about doing this with `opensearch-common` (dependency free, contains JDKJarHell) a

The thing with `xxx-common` is that is gives an impression this is something everyone needs but if we repurpose it as `JDKJarHell` - this not really the case, it becomes more like an internal utility library, may be `opensearch-bootstrap` would be better name for it?

nknize commented:
> The thing with `xxx-common` is that is gives an impression this is something everyone needs

hrm.. I viewed it like [Apache commons](https://commons.apache.org/) `Apache Commons is an Apache project focused on all aspects of reusable Java components.` Which is why I refactored the dependency free common reusable components (e.g., `Booleans`, `Nullable`, `TimeValue`) to that library.


> may be `opensearch-bootstrap` would be better name for it?

and just refactor the two jar hell classes? leave the common package in `opensearch-core`? It's not out of the question but seems like a dedicated lib is overkill? Back to why I settled on a compromise of `commons`

reta commented:
Sorry @nknize it seems like I misunderstood your suggestion:

> At minimum we'll need to create a second (dependency-free) library that provides the JarHell implementation so it can be run on the core library.

The picture I built from it: this library from above + `opensearch-core`, no `opensearch-common` for a reason that opensearch-core` + `opensearch-common` always go together. 

@andrross do you have fresh view on that? :D

nknize commented:
> ...for a reason that opensearch-core`+`opensearch-common` always go together.

Yeah, I like that justification. With the added benefit that having the `common` package in `opensearch-core` means it can be jar hell analyzed.

nknize commented:
> ...may be opensearch-bootstrap would be better name for it?

Not to bikeshed but the name should be unique to the library. `org.opensearch.bootstrap` exists in `:server`. If we split package w/ a new library of the same name then we'll need to refactor again later when we move to the jigsaw modularity framework. Maybe we should take care of that now and name it the `opensearch-build` library that is only used to build/package the project?

andrross commented:
> JarHell doesn't run on opensearch-core (because it defines the JarHell classes)

Doesn't JarHell analyze all classes in the classloader at runtime, which means it is analyzing whatever you happen to configure as your classpath? Doesn't that also inherently include the library in which JarHell is defined (because it has to be on the classpath)?

nknize commented:
> > JarHell doesn't run on opensearch-core (because it defines the JarHell classes)
> 
> Doesn't JarHell analyze all classes in the classloader at runtime, which means it is analyzing whatever you happen to configure as your classpath? Doesn't that also inherently include the library in which JarHell is defined (because it has to be on the classpath)?

JarHell [does not analyze the project the class is defined in](https://github.com/opensearch-project/OpenSearch/blob/main/buildSrc/src/main/java/org/opensearch/gradle/precommit/JarHellPrecommitPlugin.java#L50) to avoid jarHell w/ itself.

andrross commented:
> > > JarHell doesn't run on opensearch-core (because it defines the JarHell classes)
> > 
> > 
> > Doesn't JarHell analyze all classes in the classloader at runtime, which means it is analyzing whatever you happen to configure as your classpath? Doesn't that also inherently include the library in which JarHell is defined (because it has to be on the classpath)?
> 
> JarHell [does not analyze the project the class is defined in](https://github.com/opensearch-project/OpenSearch/blob/main/buildSrc/src/main/java/org/opensearch/gradle/precommit/JarHellPrecommitPlugin.java#L50) to avoid jarHell w/ itself.

So I might be the one bike-shedding here, but please bear with me :) The linked code looks to me like it just adds `:libs:opensearch-core` as a dependency to all projects except `:libs:opensearch-core` itself, but otherwise unconditionally runs the jarHell task. I did a quick experiment where I added a class (`org.junit.After`) to opensearch-core to deliberately introduce a duplicate class, then ran the following:

```
$ gd HEAD --stat
 libs/core/src/main/java/org/junit/After.java | 5 +++++
 1 file changed, 5 insertions(+)

$ ./gradlew :libs:opensearch-core:jarHell

> Task :libs:opensearch-core:jarHell FAILED
*****************************
Exec output and error:
| Output for /home/ubuntu/jdk/jdk-17/bin/java:checking for jar hell...
| java.class.path: /home/ubuntu/OpenSearch/libs/core/build/classes/java/test:/home/ubuntu/OpenSearch/libs/core/build/resources/test:/home/ubuntu/OpenSearch/libs/core/build/classes/java/main:/home/ubuntu/.gradle/caches/modules-2/files-2.1/com.google.code.findbugs/jsr305/3.0.2/25ea2e8b0c338a877313bd4672d3fe056ea78f0d/jsr305-3.0.2.jar:/home/ubuntu/OpenSearch/test/framework/build/distributions/framework-3.0.0-SNAPSHOT.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/com.carrotsearch.randomizedtesting/randomizedtesting-runner/2.7.1/e917f4983144c3b969eb7d3648338ecde5e3ba89/randomizedtesting-runner-2.7.1.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/junit/junit/4.13.2/8ac9e16d933b6fb43bc7f576336b8f4d7eb5ba12/junit-4.13.2.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.hamcrest/hamcrest/2.1/9420ba32c29217b54eebd26ff7f9234d31c3fbb2/hamcrest-2.1.jar:/home/ubuntu/OpenSearch/client/sniffer/build/distributions/opensearch-rest-client-sniffer-3.0.0-SNAPSHOT.jar:/home/ubuntu/OpenSearch/client/rest/build/distributions/opensearch-rest-client-3.0.0-SNAPSHOT.jar:/home/ubuntu/OpenSearch/libs/nio/build/distributions/opensearch-nio-3.0.0-SNAPSHOT.jar:/home/ubuntu/OpenSearch/server/build/distributions/opensearch-3.0.0-SNAPSHOT.jar:/home/ubuntu/OpenSearch/libs/cli/build/distributions/opensearch-cli-3.0.0-SNAPSHOT.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-test-framework/9.5.0-snapshot-0878271/3fb4be0744d89b5615205102bc01f4273f06fb2c/lucene-test-framework-9.5.0-snapshot-0878271.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-codecs/9.5.0-snapshot-0878271/831bf913839bc603830529366f8b489b5c0868/lucene-codecs-9.5.0-snapshot-0878271.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/commons-logging/commons-logging/1.2/4bfc12adfe4842bf07b657f0369c4cb522955686/commons-logging-1.2.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/commons-codec/commons-codec/1.15/49d94806b6e3dc933dacbd8acb0fdbab8ebd1e5d/commons-codec-1.15.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.mockito/mockito-core/4.7.0/5eeffd46d39ddd0c24666adaebc7ff62776642bd/mockito-core-4.7.0.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/net.bytebuddy/byte-buddy/1.12.18/875a9c3f29d2f6f499dfd60d76e97a343f9b1233/byte-buddy-1.12.18.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.objenesis/objenesis/3.2/7fadf57620c8b8abdf7519533e5527367cb51f09/objenesis-3.2.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.apache.httpcomponents.client5/httpclient5/5.1.4/208f9eed6d6ab709e2ae7a75b457ef60c0baefa5/httpclient5-5.1.4.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.apache.httpcomponents.core5/httpcore5/5.1.5/df9da3a1fa2351c4790245400ed28d78a8ddd3fc/httpcore5-5.1.5.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.apache.httpcomponents.core5/httpcore5-h2/5.1.5/624660339afd5006d427457e6b10b10b32fd86f1/httpcore5-h2-5.1.5.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.slf4j/slf4j-api/1.7.36/6c62681a2f655b49963a5983b8b0950a6120ae14/slf4j-api-1.7.36.jar:/home/ubuntu/OpenSearch/libs/x-content/build/distributions/opensearch-x-content-3.0.0-SNAPSHOT.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/com.fasterxml.jackson.core/jackson-core/2.14.1/7a07bc535ccf0b7f6929c4d0f2ab9b294ef7c4a3/jackson-core-2.14.1.jar:/home/ubuntu/OpenSearch/libs/secure-sm/build/distributions/opensearch-secure-sm-3.0.0-SNAPSHOT.jar:/home/ubuntu/OpenSearch/libs/geo/build/distributions/opensearch-geo-3.0.0-SNAPSHOT.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-core/9.5.0-snapshot-0878271/1f8895d068b5d98c16d48f267a772d0243148040/lucene-core-9.5.0-snapshot-0878271.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-analysis-common/9.5.0-snapshot-0878271/92c380f063c5b75efd6b3c48dffc942bfe21360f/lucene-analysis-common-9.5.0-snapshot-0878271.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-backward-codecs/9.5.0-snapshot-0878271/2f0f211b1ed0447135f84a36935625e9a33a98e5/lucene-backward-codecs-9.5.0-snapshot-0878271.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-grouping/9.5.0-snapshot-0878271/f12819e51f145624ef7b1fd93aaa543cec13e3aa/lucene-grouping-9.5.0-snapshot-0878271.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-highlighter/9.5.0-snapshot-0878271/ed88cca26580a32234e3f605cf963703ea99eb60/lucene-highlighter-9.5.0-snapshot-0878271.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-join/9.5.0-snapshot-0878271/f51934b2362e827d9f467507624b773153f0ca01/lucene-join-9.5.0-snapshot-0878271.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-memory/9.5.0-snapshot-0878271/54744097c6882a498c08ffd8a42b54da00c9420c/lucene-memory-9.5.0-snapshot-0878271.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-misc/9.5.0-snapshot-0878271/b9061dea1178e0cac86147d7b69fc53bf2f8ee58/lucene-misc-9.5.0-snapshot-0878271.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-queries/9.5.0-snapshot-0878271/eca07717c0ee563c5337222fd1a1b7ef5f03f34f/lucene-queries-9.5.0-snapshot-0878271.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-queryparser/9.5.0-snapshot-0878271/4d1ed21bc62940c4545778dc5f6249b67d08b095/lucene-queryparser-9.5.0-snapshot-0878271.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-sandbox/9.5.0-snapshot-0878271/c564c73080cff226a131847b4361bc9c2155f00c/lucene-sandbox-9.5.0-snapshot-0878271.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-spatial-extras/9.5.0-snapshot-0878271/7a654fcc671e0372a9becd7c7cbd9a1f43569106/lucene-spatial-extras-9.5.0-snapshot-0878271.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-spatial3d/9.5.0-snapshot-0878271/33c9c43eac6b4924155eeff03d2231c2bed9b169/lucene-spatial3d-9.5.0-snapshot-0878271.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-suggest/9.5.0-snapshot-0878271/95a69fd51377d964d16057a4a9501665a2ca9a7a/lucene-suggest-9.5.0-snapshot-0878271.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/com.carrotsearch/hppc/0.8.1/ffc7ba8f289428b9508ab484b8001dea944ae603/hppc-0.8.1.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/joda-time/joda-time/2.10.13/86f338c18cea2a89005556642e81707ff920dd38/joda-time-2.10.13.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/com.tdunning/t-digest/3.3/5e96c4fd7d63b05828cf5ef41da20649195b1b78/t-digest-3.3.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.hdrhistogram/HdrHistogram/2.1.12/6eb7552156e0d517ae80cc2247be1427c8d90452/HdrHistogram-2.1.12.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.locationtech.spatial4j/spatial4j/0.7/faa8ba85d503da4ab872d17ba8c00da0098ab2f2/spatial4j-0.7.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.locationtech.jts/jts-core/1.15.0/705981b7e25d05a76a3654e597dab6ba423eb79e/jts-core-1.15.0.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.apache.logging.log4j/log4j-api/2.17.1/d771af8e336e372fb5399c99edabe0919aeaf5b2/log4j-api-2.17.1.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.apache.logging.log4j/log4j-jul/2.17.1/881333b463d47828eda7443b19811763367b1916/log4j-jul-2.17.1.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.apache.logging.log4j/log4j-core/2.17.1/779f60f3844dadc3ef597976fcb1e5127b1f343d/log4j-core-2.17.1.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/net.java.dev.jna/jna/5.5.0/e0845217c4907822403912ad6828d8e0b256208/jna-5.5.0.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/net.sf.jopt-simple/jopt-simple/5.0.4/4fdac2fbe92dfad86aa6e9301736f6b4342a3f5c/jopt-simple-5.0.4.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/org.yaml/snakeyaml/1.32/e80612549feb5c9191c498de628c1aa80693cf0b/snakeyaml-1.32.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/com.fasterxml.jackson.dataformat/jackson-dataformat-smile/2.14.1/656ccecc1fc85b95d13e5b8080289fc1a5e5e21e/jackson-dataformat-smile-2.14.1.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/com.fasterxml.jackson.dataformat/jackson-dataformat-yaml/2.14.1/cf6d18651659a2e64301452c841e6daa62e77bf6/jackson-dataformat-yaml-2.14.1.jar:/home/ubuntu/.gradle/caches/modules-2/files-2.1/com.fasterxml.jackson.dataformat/jackson-dataformat-cbor/2.14.1/4e6fbcdcd2a01e4a5cb5901338cab6199c9b26b/jackson-dataformat-cbor-2.14.1.jar
| sun.boot.class.path: null
| java.home: /home/ubuntu/jdk/jdk-17
| examining directory: /home/ubuntu/OpenSearch/libs/core/build/classes/java/test
| examining directory: /home/ubuntu/OpenSearch/libs/core/build/resources/test
| examining directory: /home/ubuntu/OpenSearch/libs/core/build/classes/java/main
| examining jar: /home/ubuntu/.gradle/caches/modules-2/files-2.1/com.google.code.findbugs/jsr305/3.0.2/25ea2e8b0c338a877313bd4672d3fe056ea78f0d/jsr305-3.0.2.jar
| examining jar: /home/ubuntu/OpenSearch/test/framework/build/distributions/framework-3.0.0-SNAPSHOT.jar
| examining jar: /home/ubuntu/.gradle/caches/modules-2/files-2.1/com.carrotsearch.randomizedtesting/randomizedtesting-runner/2.7.1/e917f4983144c3b969eb7d3648338ecde5e3ba89/randomizedtesting-runner-2.7.1.jar
| examining jar: /home/ubuntu/.gradle/caches/modules-2/files-2.1/junit/junit/4.13.2/8ac9e16d933b6fb43bc7f576336b8f4d7eb5ba12/junit-4.13.2.jar
| Exception in thread "main" java.lang.IllegalStateException: jar hell!
| class: org.junit.After
| jar1: /home/ubuntu/OpenSearch/libs/core/build/classes/java/main
| jar2: /home/ubuntu/.gradle/caches/modules-2/files-2.1/junit/junit/4.13.2/8ac9e16d933b6fb43bc7f576336b8f4d7eb5ba12/junit-4.13.2.jar
|       at org.opensearch.bootstrap.JarHell.checkClass(JarHell.java:314)
|       at org.opensearch.bootstrap.JarHell.checkJarHell(JarHell.java:213)
|       at org.opensearch.bootstrap.JarHell.checkJarHell(JarHell.java:100)
|       at org.opensearch.bootstrap.JarHell.main(JarHell.java:84)

FAILURE: Build failed with an exception.
```

Am I missing something? It looks to me that we're running the JarHell task on the library that JarHell itself is defined in. Assuming that's the case, then I would lean towards not having a separate commons library for the reason @reta cited that commons and core always go together.

nknize commented:
Just getting back to this..

> ...commons and core always go together.

The move to refactor large `:server` components to core means `:libs::opensearch-core` depends on `:libs:opensearch-commons` but not the other way around. This affects projects like `:buildtools` that currently depends on `:libs:opensearch-core` but not `:server`. In the refactor effort `:libs:opensearch-core` will be pulling in dependencies not needed for `:buildtools` (e.g., lucene, log4j). We need a library like `:opensearch-commons` for classes that are dependency free for any other libraries that need foundation classes (e.g., `PathUtils`) without pulling in the monolith of `:libs:core` (currently `:server`).

reta commented:
> Just getting back to this..
> 
> > ...commons and core always go together.
> 
> The move to refactor large `:server` components to core means `:libs::opensearch-core` depends on `:libs:opensearch-commons` but not the other way around. This affects projects like `:buildtools` that currently depends on `:libs:opensearch-core` but not `:server`. In the refactor effort `:libs:opensearch-core` will be pulling in dependencies not needed for `:buildtools` (e.g., lucene, log4j). We need a library like `:opensearch-commons` for classes that are dependency free for any other libraries that need foundation classes (e.g., `PathUtils`) without pulling in the monolith of `:libs:core` (currently `:server`).

@nknize hm ... do you mean this dependency from `:buildtools`?

```
integTestRuntimeOnly(project(":libs:opensearch-core"))
```

nknize commented:
> ... do you mean this dependency from `:buildtools`?

yes. And [`:libs:opensearch-cli`](https://github.com/opensearch-project/OpenSearch/blob/main/libs/cli/build.gradle#L36) and [`:libs:opensearch-nio`](https://github.com/opensearch-project/OpenSearch/blob/main/libs/nio/build.gradle), ssl-config, etc. These should depend on `:libs:opensearch-common` and clear the way for `:libs:opensearch-core` to hold `:server` interfaces, and core components. 

I'm also refactoring the core library to the `o.opensearch.core.` namespace, the `.common` llibrary to the `o.opensearch.common`, to fix the split package requirement in prep for jigsaw modularization. That's what these incremental PRs are all about.

nknize commented:
Oops... I pushed a hard revert. Thank the gods for reflog recovery :) ... reopeneing

nknize commented:
This is ready to go. I renamed the library to `:libs:opensearch-common`, added javadoc checks, and refactored jarhell from core. `opensearch-core` is leaned out and, for now, the namespace is the same as before so we can backport to 2.x. After this merges I'll fix up the xcontent refactor PR to retain the namespace for a 2.x backport and start opening follow on PRs to incrementally refactor `:server` interfaces and foundation components to core

reta commented on buildSrc/build.gradle:
```diff
@@ -191,6 +191,7 @@ if (project != rootProject) {
     distribution project(':distribution:archives:linux-tar')
     distribution project(':distribution:archives:windows-zip')
 
+    integTestRuntimeOnly(project(":libs:opensearch-common"))
```
Sorry @nknize, referring to your previous comment:

> In the refactor effort :libs:opensearch-core will be pulling in dependencies not needed for :buildtools (e.g., lucene, log4j). We need a library like :opensearch-commons for classes that are dependency free for any other libraries that need foundation classes (e.g., PathUtils) without pulling in the monolith of :libs:core (currently :server).

But I see we still need both, why is that? I expected to `integTestRuntimeOnly(project(":libs:opensearch-common"))` only, what is missed to make it happen?

nknize replied:
Because the remaining `internal` classes in core are still needed by `build-tools`. I opted not to bring them over in this PR and instead move them in a follow up after the refactoring x-content PR. Really for no other reason than to avoid having an empty project.

reta commented on buildSrc/src/main/java/org/opensearch/gradle/precommit/JarHellPrecommitPlugin.java:
```diff
@@ -44,11 +44,11 @@ public class JarHellPrecommitPlugin extends PrecommitPlugin {
             // Internal projects are not all plugins, so make sure the check is available
             // we are not doing this for this project itself to avoid jar hell with itself
-            project.getDependencies().add("jarHell", project.project(":libs:opensearch-core"));
+            project.getDependencies().add("jarHell", project.project(":libs:opensearch-common"));
```
👍

reta commented on libs/common/build.gradle:
```diff
@@ -0,0 +1,35 @@
+
+apply plugin: 'opensearch.publish'
+
+archivesBaseName = 'opensearch-common'
```
👍

reta commented on libs/common/build.gradle:
```diff
@@ -0,0 +1,35 @@
+  testImplementation "junit:junit:${versions.junit}"
+  testImplementation "org.hamcrest:hamcrest:${versions.hamcrest}"
+
+  testImplementation(project(":test:framework")) {
```
The `:test:framework` brings whole bucket of other dependencies (even for tests only), why do we need it?

nknize replied:
Because the tests that have been refactored from core extend `OpenSearchTestCase` which needs to test framework. These are only used in the verification tests and aren't packaged in the distribution jar

reta commented on libs/common/src/main/java/org/opensearch/common/collect/Map.java:
```diff
@@ -32,6 +32,13 @@
+/**
+ * Java 9 Map
+ *
+ * todo: deprecate and remove w/ min jdk upgrade to 11?
```
Could we do that since 3.0.0 has JDK-11 baseline?

nknize replied:
Yes. I was planning it in a follow up PR.

reta commented:
I think LGTM, @andrross what do you think? A side note, @nknize could you please update https://github.com/opensearch-project/OpenSearch/issues/5910, I believe the presence of `opensearch-common` is more or less justified, but not formalized, thank you.

nknize commented:
> ...could you please update https://github.com/opensearch-project/OpenSearch/issues/5910

[Done](https://github.com/opensearch-project/OpenSearch/issues/5910#issuecomment-1408813473).

andrross commented on CHANGELOG.md:
```diff
@@ -46,6 +46,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - Changed http code on create index API with bad input raising NotXContentException from 500 to 400 ([#4773](https://github.com/opensearch-project/OpenSearch/pull/4773))
 - Change http code for DecommissioningFailedException from 500 to 400 ([#5283](https://github.com/opensearch-project/OpenSearch/pull/5283))
 - Require MediaType in Strings.toString API ([#6009](https://github.com/opensearch-project/OpenSearch/pull/6009))
+- [Refactor] core.common to new opensearch-common library ([#5976](https://github.com/opensearch-project/OpenSearch/pull/5976))
```
This needs to go in `[Unreleased 2.x]` assuming it will be backported

andrross commented on libs/common/src/main/java/org/opensearch/common/collect/List.java:
```diff
@@ -34,6 +34,13 @@
+ *
+ * todo: deprecate and remove w/ min jdk upgrade to 11?
+ *
+ * @opensearch.internal
```
This is a nitpick, but I think this is a bit of an abuse of the `@opensearch.internal` annotation. This was clearly meant to be a public utility just like anything else in the common namespace. It just has been superseded by an alternative, which is exactly what the `java.lang.Deprecated` annotation is meant to convey.
